beluga_ros/tools/conversion_utils.py

```diff
@@ -0,0 +1,273 @@
+        assert yaml_path.is_file(), f"file {yaml_path} doesn't exist."
+        assert yaml_path.suffix == ".yaml"
+
+        with open(yaml_path, "rb") as fp:
+            data = yaml.safe_load(fp)
+
+        pgm_path: Path = yaml_path.parent / data["image"]
+        assert pgm_path.is_file(), f".pgm file at {pgm_path} does not exist"
+
+        with open(yaml_path.parent / data["image"], "rb") as pgmf:
+            grid = plt.imread(pgmf)
+
+        return OccupancyGrid(
+            resolution=data["resolution"], origin=np.asarray(data["origin"]), grid=grid
+        )
+
+
+def grid_to_point_cloud(occupancy_grid: OccupancyGrid) -> np.ndarray:
+    """
+    Convert an OccupancyGrid's occupied cells to a 2D point cloud.
+
+    Uses the center of the cell for the conversion to reduce max error.
+    """
+    occupied_cells_indices = np.asarray(
+        np.where(occupancy_grid.grid == 0)
+    )  # in ROS maps 0 == occupied.
+    res = occupancy_grid.resolution
+
+    # Discretized occupied cells using the center of the cell.
+    points = occupied_cells_indices * res + np.ones_like(occupied_cells_indices) * (
+        res / 2
+    )
+    # Compensate for origin
+    points[0] += occupancy_grid.origin[0]
+    points[1] += occupancy_grid.origin[1]
+    return points
+
+
+def fit_normal_distribution(points: np.ndarray) -> Optional[NormalDistribution]:
+    """Fit a normal distribution to a set of 2D points."""
+    assert points.shape[0] == 2
+    # Literature suggests doing this check to avoid singularities.
+    # See The Three-Dimensional Normal-Distributions Transform– an Efficient
+    # Representation for Registration, Surface Analysis, and Loop Detection
+    # by Martin Magnusson, 2009 chapter 6.
+    if points.shape[1] <= 5:
+        return None
+    mean = points.mean(axis=1)
+    covariance = np.cov(points)
+
+    # avoid singularities by enforcing a minimum variance in both dimensions.
+    covariance[0, 0] = max(covariance[0, 0], 5e-3)
+    covariance[1, 1] = max(covariance[1, 1], 5e-3)
+
+    return NormalDistribution(mean=mean, covariance=covariance)
+
+
+def point_cloud_to_ndt(pc: np.ndarray, cell_size=1.0) -> NDTMap:
+    """
+    Convert a 2D point cloud into a NDT map representation.
+
+    Does so by clustering points in 2D cells of {cell_size} * {cell_size} meters^2
+    and fitting a normal distribution when applicable.
+    """
+    assert pc.shape[0] == 2
+    points_clusters: Dict[DiscreteCell, np.ndarray] = {}
+    discretized_points = np.floor(pc / cell_size).astype(np.int64)
+    cells = np.unique(discretized_points, axis=-1).T
+
+    for cell in cells:
+        pts_in_cell = np.all(discretized_points.T == cell, axis=1)
+        points_clusters[DiscreteCell(x=cell[0], y=cell[1])] = pc[:, pts_in_cell]
+
+    ret = NDTMap(resolution=cell_size)
+
+    for cell, points in points_clusters.items():
+        if (dist := fit_normal_distribution(points)) is not None:
```

Comment:
@serraramiro1 ROS 2 Humble requires Python 3.6+ (see [REP-2000](https://www.ros.org/reps/rep-2000.html#humble-hawksbill-may-2022-may-2027)). Assignments on `if` statements were not introduced until Python 3.8. Consider:
```suggestion
        dist = fit_normal_distribution(points)
        if dist is not None:
```
instead.

Reply:
Done